pkg/controller/drivelvgcontroller/drivelvgcontroller.go

```diff
@@ -0,0 +1,365 @@
+		Complete(d)
+}
+
+// Reconcile reconciles Drive custom resources
+func (d *Controller) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	defer metricsC.ReconcileDuration.EvaluateDurationForType("csicontroller_drive_controller")()
+	resourceName := req.Name
+	ctx, cancelFn := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancelFn()
+
+	log := d.log.WithFields(logrus.Fields{"method": "Reconcile", "name": resourceName})
+
+	drive := &drivecrd.Drive{}
+	if err := d.client.ReadCR(ctx, resourceName, "", drive); err == nil {
+		return d.reconcileDrive(ctx, drive)
+	}
+	log.Warnf("Failed to read Drive %s CR, try to read LVG CR", resourceName)
+	lvg := &lvgcrd.LogicalVolumeGroup{}
+	if err := d.client.ReadCR(ctx, resourceName, "", lvg); err != nil {
+		log.Errorf("Failed to read LVG %s CR", resourceName)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	return d.reconcileLVG(lvg)
+}
+
+// reconcileLVG perform logic for LVG reconciliation
+func (d *Controller) reconcileLVG(lvg *lvgcrd.LogicalVolumeGroup) (ctrl.Result, error) {
+	log := d.log.WithFields(logrus.Fields{
+		"method": "reconcileLVG",
+	})
+	var (
+		status = lvg.Spec.GetStatus()
+		health = lvg.Spec.GetHealth()
+		name   = lvg.GetName()
+	)
+	// If LVG status is failed or Health is not good, try to reset its AC size to 0
+	if status == apiV1.Failed || health != apiV1.HealthGood {
+		return ctrl.Result{}, d.resetACSizeOfLVG(name)
+	}
+	// If LVG is already presented on a machine but doesn't have AC, try to create its AC using annotation with
+	// VG free space
+	size, err := getFreeSpaceFromLVGAnnotation(lvg.Annotations)
+	if err != nil {
+		log.Errorf("Failed to get free space from LVG %v, err: %v", lvg, err)
+		if err == errTypes.ErrorNotFound {
+			err = nil
+		}
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, d.createLVGACIfNotExistsOrUpdate(lvg, size)
+}
+
+// reconcileDrive preforms logic for drive reconciliation
+func (d *Controller) reconcileDrive(ctx context.Context, drive *drivecrd.Drive) (ctrl.Result, error) {
+	var (
+		health  = drive.Spec.GetHealth()
+		status  = drive.Spec.GetStatus()
+		isClean = drive.Spec.GetIsClean()
+	)
+	switch {
+	case isClean && (health == apiV1.HealthGood && status == apiV1.DriveStatusOnline):
+		return d.createACIfNotExistOrUpdate(ctx, drive.Spec)
+	case health != apiV1.HealthGood || status != apiV1.DriveStatusOnline:
+		return d.handleDriveIsNotGood(ctx, drive.Spec)
+	case !isClean:
+		return d.createACIfNotExistOrUpdate(ctx, drive.Spec)
+	}
+	return ctrl.Result{RequeueAfter: time.Minute}, nil
+}
+
+// createACIfNotExistOrUpdate tries to create AC for drive or update its size if AC already exists
+func (d *Controller) createACIfNotExistOrUpdate(ctx context.Context, drive api.Drive) (ctrl.Result, error) {
+	log := d.log.WithFields(logrus.Fields{
+		"method": "createACIfNotExistOrUpdate",
+	})
+	driveUUID := drive.GetUUID()
+	size := drive.GetSize()
+	// if drive is not clean, size is 0
+	if !drive.GetIsClean() {
```

Comment:
Won't work for LVG

Reply:
Could you please add more details and information about the problem?